java/vector/src/main/java/org/apache/arrow/vector/file/ArrowWriter.java

```diff
@@ -0,0 +1,179 @@
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.arrow.vector.schema.ArrowBuffer;
+import org.apache.arrow.vector.schema.ArrowRecordBatch;
+import org.apache.arrow.vector.schema.FBSerializable;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.flatbuffers.FlatBufferBuilder;
+
+import io.netty.buffer.ArrowBuf;
+
+public class ArrowWriter implements AutoCloseable {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ArrowWriter.class);
+
+  private static final byte[] MAGIC = "ARROW1".getBytes();
+
+  private final WritableByteChannel out;
+
+  private final Schema schema;
+
+  private final List<ArrowBlock> recordBatches = new ArrayList<>();
+
+  private long currentPosition = 0;
+
+  private boolean started = false;
+
+  public ArrowWriter(WritableByteChannel out, Schema schema) {
+    this.out = out;
+    this.schema = schema;
+  }
+
+  private void start() throws IOException {
+    writeMagic();
+  }
+
+  private long write(byte[] buffer) throws IOException {
+    return write(ByteBuffer.wrap(buffer));
+  }
+
+  private long writeZeros(int zeroCount) throws IOException {
+    return write(new byte[zeroCount]);
+  }
+
+  private long align() throws IOException {
+    if (currentPosition % 8 != 0) { // align on 8 byte boundaries
+      return writeZeros(8 - (int)(currentPosition % 8));
+    }
+    return 0;
+  }
+
+  private long write(ByteBuffer buffer) throws IOException {
+    long length = buffer.remaining();
+    out.write(buffer);
+    currentPosition += length;
+    return length;
+  }
+
+  private static byte[] intToBytes(int value) {
+    byte[] outBuffer = new byte[4];
+    outBuffer[3] = (byte)(value >>> 24);
+    outBuffer[2] = (byte)(value >>> 16);
+    outBuffer[1] = (byte)(value >>>  8);
+    outBuffer[0] = (byte)(value >>>  0);
+    return outBuffer;
+  }
+
+  private long writeIntLittleEndian(int v) throws IOException {
+    return write(intToBytes(v));
+  }
+
+  // TODO: write dictionaries
+
+  public void writeRecordBatch(ArrowRecordBatch recordBatch) throws IOException {
+    checkStarted();
+    align();
+    // write metadata header
+    long offset = currentPosition;
+    write(recordBatch);
+    align();
+    // write body
+    long bodyOffset = currentPosition;
+    List<ArrowBuf> buffers = recordBatch.getBuffers();
+    List<ArrowBuffer> buffersLayout = recordBatch.getBuffersLayout();
+    if (buffers.size() != buffersLayout.size()) {
+      throw new IllegalStateException("the layout does not match: " + buffers.size() + " != " + buffersLayout.size());
+    }
+    for (int i = 0; i < buffers.size(); i++) {
+      ArrowBuf buffer = buffers.get(i);
+      ArrowBuffer layout = buffersLayout.get(i);
+      long startPosition = bodyOffset + layout.getOffset();
+      if (startPosition != currentPosition) {
+        writeZeros((int)(startPosition - currentPosition));
+      }
+      write(buffer);
```

Comment:
Actually ArrowRecordBatch just pads in between the buffers if they are not a multiple of 8.
Same Thing in the ArrowWriter. The buffers are written on 8-byte alignments. There can be gaps in between.